server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -607,21 +608,49 @@ void processBulkIndexIngestRequest(Task task, BulkRequest original, ActionListen
                         // (this will happen if pre-processing all items in the bulk failed)
                         actionListener.onResponse(new BulkResponse(new BulkItemResponse[0], 0));
                     } else {
-                        doExecute(task, bulkRequest, actionListener);
+                        // If a processor went async and returned a response on a different thread then
+                        // before we continue the bulk request we should fork back on a write thread:
+                        if (Thread.currentThread().getName().contains(ThreadPool.Names.WRITE)) {
+                            doExecute(task, bulkRequest, actionListener);
+                        } else {
+                            threadPool.executor(ThreadPool.Names.WRITE).execute(new AbstractRunnable() {
+                                @Override
+                                public void onFailure(Exception e) {
+                                    listener.onFailure(e);
+                                }
+
+                                @Override
+                                protected void doRun() throws Exception {
+                                    doExecute(task, bulkRequest, actionListener);
+                                }
+
+                                @Override
+                                public boolean isForceExecution() {
+                                    // If we fork back to a write thread we should fail, because tp queue is full.
```

Comment:
// If we fork back to a write thread we should ?not? fail
or
// If we fork back to a write thread we ?can? fail
(or am i mis-understanding)

Reply:
whoops I meant: If we fork back to a write thread we should **not** fail